text/0244-eks-developer-preview-migration-path.md

```diff
@@ -0,0 +1,102 @@
+---
+feature name: Migration Path For EKS Developer Preview
+start date: 2020-08-26
+rfc pr: https://github.com/aws/aws-cdk-rfcs/pull/245
+related issue: https://github.com/aws/aws-cdk-rfcs/issues/244
+---
+
+# Summary
+
+This RFC proposes a migration path for existing users following the upcoming transition of the `aws-eks` module from *Experimental* to *Developer Preview*.
+
+# Motivation
+
+As part of transitioning the `aws-eks` module to *Developer Preview*, we need to introduce a few breaking changes that require
```

Comment:
I'm pretty interested in the specifics of this case. Breaking  changes and replacement don't *have* to coincide. Can you go a little into the changes that prompted this RFC? Have we already done the work to see if we can avoid this process completely?

Reply:
> Breaking changes and replacement don't have to coincide

You're right, and actually the changes we talk about are not breaking the API. But since they require cluster replacement, we treat them with even more care than API breakage.

> Can you go a little into the changes that prompted this RFC?

Essentially we weren't using the right mechanisms to configure the cluster `connections` and its interaction with our custom resource that runs `kubectl` commands against it. To make it work properly, we need to change the VPC configuration of the cluster, which requires replacement.

I'm not sure if you wanted more deep technical details, if so let me know, happy to talk about it. 

You can have a look at this [issue](https://github.com/aws/aws-cdk/issues/9754), which kind of prompted this whole thing. 

> Have we already done the work to see if we can avoid this process completely?

We have. Strictly speaking, we can probably get away without requiring cluster replacement. However, as we dove deeper into how EKS works, we realized that the our current configuration is messy, fragile, and in some (albeit rare) cases, is just plain wrong. In addition, we already had a feature that required cluster replacement a while ago, and eventually managed to avoid it at the expense of keeping a complicated stack modeling and introducing a **1 cluster per stack** limitation. 

At the end of the day, we decided we should do the right thing for GA, and make the cluster as robust as possible so that we minimize the change of having to do cluster replacements post GA.